modelforge/dataset/dataset.py

```diff
@@ -476,67 +508,265 @@ def __init__(
         split: SplittingStrategy = RandomRecordSplittingStrategy(),
         batch_size: int = 64,
         split_file: Optional[str] = None,
+        transform: Optional[nn.Module] = None,
     ):
         super().__init__()
         self.data = data
         self.batch_size = batch_size
-        self.split_idxs: Optional[str] = None
-        self.train_dataset = Optional[TorchDataset]
-        self.test_dataset = Optional[TorchDataset]
-        self.val_dataset = Optional[TorchDataset]
+        self.split_idxs = None
+        self.train_dataset = None
+        self.test_dataset = None
+        self.val_dataset = None
+        self.transform = transform
+        self.split = split
+        self.split_file = split_file
+        self.dataset_statistics = {}
+
+    @classmethod
+    def calculate_self_energies(
+        cls, dataset: TorchDataset, collate: bool = True
+    ) -> Dict[str, float]:
+        """
+        Calculates the self energies for each atomic number in the dataset by performing a least squares regression.
 
-        if split_file:
-            import numpy as np
+        Parameters
+        ----------
+        dataset : TorchDataset
+            The dataset from which to calculate self energies.
+        collate : bool, optional
+            If True, uses a custom collate function to gather batch data. Defaults to True.
 
-            logger.debug(f"Loading split indices from {split_file}")
-            self.split_idxs = np.load(split_file)
-        else:
-            logger.debug(f"Using splitting strategy {split}")
-            self.split = split
+        Returns
+        -------
+        Dict[int, float]
+            A dictionary mapping atomic numbers to their calculated self energies.
+        """
+        log.info("Computing self energies for elements in the dataset.")
+
+        # Define the collate function based on the `collate` parameter
+        collate_fn = collate_conformers if collate else None
+
+        # Initialize variables to hold data for regression
+        batch_size = 64
+        # Determine the size of the counts tensor
+        num_molecules = dataset.n_records
+        # Determine up to which Z we detect elements
+        max_atomic_number = 100
```

Comment:
I agree that this might be a source of future complications. One way to solve this is to calculate max Z from the dataset that is loaded.

Reply:
Sorry, let me clarify this comment: I don't mean the torch dataset, I mean the internal representation in the h5 file.

Reply:
I guess the complication of having max_atomic_number be a variable is that a user could under-allocate (we would need to do a check as we read in the batches to make sure haven't under allocated).  It is probably fine to just leave it as is, since I doubt we are going to be trending into the synthetic elements anytime soon, for obvious reasons.  And again, since this didn't seem to in anyway impact the speed of the calculation, I think it's fine to just pad out to 100 for safety.